clients/application/src/main/java/org/eclipse/hono/application/client/ApplicationClientFactory.java

```diff
@@ -79,4 +79,16 @@
 
     // TODO add methods for command & control
 
+    /**
+     * Creates a client for sending commands to devices of the given tenant asynchronously, i.e. command responses
+     * get received by a separate receiver.
+     *
+     * @param tenantId The tenant of the devices to which commands shall be sent.
+     * @return A future that will complete with the command client (if successful) or
+     *         fail if the client cannot be created, e.g. because the underlying connection
+     *         is not established or if a concurrent request to create a client for the same
+     *         tenant is already being executed.
+     * @throws NullPointerException if tenantId is {@code null}.
+     */
+    Future<AsyncCommandClient> createAsyncCommandClient(String tenantId);
```

Comment:
I do not quite understand why we need to create another object here. Can't we simply let the *ApplicationClientFactory* implement *AsyncCommandClient*? One of the goals of the new client was to no longer create tenant specific instances of clients but instead allow users to simply pass in a tenant ID to the commands directly instead. Or is there a particular reason that would require use to use the old approach here?

Reply:
> Can't we simply let the _ApplicationClientFactory_ implement _AsyncCommandClient_?

With implement do you mean  _ApplicationClientFactory_ to have directly  `sendAsyncCommand(...)`. FMPOV then this approach doesn't approach the name _ApplicationClientFactory_ as it is a factory for creating clients.


> One of the goals of the new client was to no longer create tenant specific instances of clients but instead allow users to simply pass in a tenant ID to the commands directly instead.

I understand the intention now. In this case, instead of `Future<AsyncCommandClient> createAsyncCommandClient(String tenantId)` I would change it to `Future<AsyncCommandClient> createAsyncCommandClient()`.

Reply:
> I understand the intention now. In this case, instead of Future<AsyncCommandClient> createAsyncCommandClient(String tenantId) I would change it to Future<AsyncCommandClient> getOrCreateAsyncCommandClient().

FMPOV there is no need for a factory method here.

> FMPOV then this approach doesn't approach the name ApplicationClientFactory as it is a factory for creating clients.

I agree, we should rename it to just *ApplicationClient* :-)

Reply:
> I agree, we should rename it to just _ApplicationClient_ :-)

:+1:

`ApplicationClient` can extend only one class. In that case, I would group `CommandClient` and `AsyncCommandClient` methods under `CommandSender`. Then the `ApplicationClient` can extend `CommandSender`

Reply:
@sophokles73 I have renamed now to `ApplicationClient`. Also the `ApplicationClient` now implements the `CommandSender`. Would you mind taking a look?